src/addons/messagelog/src/main/java/ee/ria/xroad/proxy/messagelog/LogArchiver.java

```diff
@@ -195,14 +207,24 @@ private Path getWorkingPath() {
 
     @SuppressWarnings("unchecked")
     protected List<TimestampRecord> getNonArchivedTimestampRecords(
-            Session session, int maxRecordsToGet) {
+            Session session, int maxRecordsToGet, long maxTimestampId) {
         Criteria criteria = session.createCriteria(TimestampRecord.class);
         criteria.add(Restrictions.eq("archived", false));
+        criteria.add(Restrictions.le("id", maxTimestampId));
         criteria.setMaxResults(maxRecordsToGet);
+        criteria.addOrder(Order.asc("id"));
         return criteria.list();
     }
 
     @SuppressWarnings("unchecked")
+    protected long getMaxTimestampId(
+            Session session) {
+        Criteria criteria = session.createCriteria(TimestampRecord.class);
+        criteria.setProjection(Projections.max("id"));
+        return (long)criteria.uniqueResult();
```

Comment:
Should include 'archived=false' criterion so that the partial index can be used.